homeassistant/components/eheimdigital/coordinator.py

```diff
@@ -0,0 +1,79 @@
+"""Data update coordinator for the EHEIM Digital integration."""
+
+from __future__ import annotations
+
+from collections.abc import Callable, Coroutine
+from typing import Any
+
+from eheimdigital.device import EheimDigitalDevice
+from eheimdigital.hub import EheimDigitalHub
+from eheimdigital.types import EheimDeviceType
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.entity_component import DEFAULT_SCAN_INTERVAL
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN, LOGGER
+
+type AsyncSetupDeviceEntitiesCallback = Callable[[str], Coroutine[Any, Any, None]]
+
+
+class EheimDigitalUpdateCoordinator(
+    DataUpdateCoordinator[dict[str, EheimDigitalDevice]]
+):
+    """The EHEIM Digital data update coordinator."""
+
+    platform_callbacks: dict[Platform, AsyncSetupDeviceEntitiesCallback]
+    config_entry: ConfigEntry
+    hub: EheimDigitalHub
+    known_devices: set[str]
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the EHEIM Digital data update coordinator."""
+        super().__init__(
+            hass, LOGGER, name=DOMAIN, update_interval=DEFAULT_SCAN_INTERVAL
+        )
+        self.hub = EheimDigitalHub(
+            host=self.config_entry.data[CONF_HOST],
+            session=async_get_clientsession(hass),
+            loop=hass.loop,
+            receive_callback=self._async_receive_callback,
+            device_found_callback=self._async_device_found,
+        )
+        self.known_devices = set()
+        self.platform_callbacks = {}
+
+    def add_platform_callback(
+        self,
+        platform: Platform,
+        async_setup_device_entities: AsyncSetupDeviceEntitiesCallback,
+    ) -> None:
+        """Add the setup callbacks from a specific platform."""
+        if platform not in self.platform_callbacks:
+            self.platform_callbacks[platform] = async_setup_device_entities
+
+    async def _async_device_found(
+        self, device_address: str, device_type: EheimDeviceType
+    ) -> None:
+        """Set up a new device found.
+
+        This function is called from the library whenever a new device is added.
+        """
+
+        if device_address not in self.known_devices:
+            for platform_callback in self.platform_callbacks.values():
+                await platform_callback(device_address)
```

Comment:
This is much better indeed. I am wondering what the significance of storing it as a `dict` with `platform` as key, as that is not used, over just keeping it in a set or list and just going over the list.

Reply:
Oh yeah, forgot to simplify further, thanks!